homeassistant/components/ipp/config_flow.py

```diff
@@ -0,0 +1,171 @@
+"""Config flow to configure the IPP integration."""
+import logging
+from typing import Any, Dict, Optional
+
+from pyipp import IPP, IPPConnectionError, IPPConnectionUpgradeRequired
+import voluptuous as vol
+
+from homeassistant.config_entries import (
+    CONN_CLASS_LOCAL_POLL,
+    SOURCE_ZEROCONF,
+    ConfigFlow,
+)
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PORT,
+    CONF_SSL,
+    CONF_VERIFY_SSL,
+)
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.typing import ConfigType
+
+from .const import CONF_BASE_PATH, CONF_UUID
+from .const import DOMAIN  # pylint: disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class IPPFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle an IPP config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_LOCAL_POLL
+
+    async def async_step_user(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initiated by the user."""
+        return await self._handle_config_flow(user_input)
+
+    async def async_step_zeroconf(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Handle zeroconf discovery."""
+        if user_input is None:
+            return self.async_abort(reason="connection_error")
+
+        zeroconf_type = user_input["type"]
+        # Hostname is format: EPSON123456.local.
+        host = user_input["hostname"].rstrip(".")
+        port = user_input["port"]
+        name, _ = host.rsplit(".")
+        ipp_rp = user_input["properties"].get("rp", "ipp/printer")
+        tls = zeroconf_type == "_ipps._tcp.local."
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context.update(
+            {
+                CONF_HOST: host,
+                CONF_PORT: port,
+                CONF_SSL: tls,
+                CONF_VERIFY_SSL: False,
+                CONF_BASE_PATH: "/" + ipp_rp,
+                CONF_NAME: name,
+                CONF_UUID: user_input["properties"].get("UUID"),
+                "title_placeholders": {"name": name},
+            }
+        )
+
+        # Prepare configuration flow
+        return await self._handle_config_flow(user_input, True)
+
+    async def async_step_zeroconf_confirm(
+        self, user_input: ConfigType = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initiated by zeroconf."""
+        return await self._handle_config_flow(user_input)
+
+    async def _handle_config_flow(
```

Comment:
I feel like things have gotten more complicated by trying to create reusable code .

Here are some things that are missing / can be improved:
 - Zeroconf needs to set UUID and abort before returning confirm form
 - Return the form from the step function that is going to handle it. So inside `async_step_zeroconf_confirm`, check if `user_input is None` and return form if so. Then in `step_zeroconf` you do `return await self.async_step_zeroconf_confirm()`.
 - Don't store data in `context` but store it on the instance
 - I would greate a helper function that instantiates an `IPP` object, verifies it can connect, does not catch exceptions, and then returns the uuid. This is similar to how [our config flow scaffolding works](https://github.com/home-assistant/core/blob/dev/script/scaffold/templates/config_flow/integration/config_flow.py#L31). Handle the exceptions in the steps and return appropriate form/abort.

Reply:
This config flows follows how wled does an I believe does cover the cases you mentioned with zeroconf. I guess it could be rewritten but not sure why wled can do it this way and not this integration?

Reply:
Just because it was missed in 1 integration cannot be used as an argument to do it like that in other integrations. The code is really hard to follow now and I don't think that's good for maintenance.

Reply:
fair enough :) I guess I was considering frencks work to be the ideal basis. I've already started adjusting